lib/tran_sock.c

```diff
@@ -115,10 +115,15 @@ MOCK_DEFINE(tran_sock_send_iovec)(int sock, uint16_t msg_id, bool is_reply,
         memcpy(CMSG_DATA(cmsg), fds, size);
     }
 
-    // FIXME: this doesn't check the entire data was sent?
-    ret = sendmsg(sock, &msg, 0);
+    ret = sendmsg(sock, &msg, MSG_NOSIGNAL);
     if (ret == -1) {
+        /* Treat a failed write due to EPIPE the same as a short write. */
+        if (errno == EPIPE) {
+            return -ECONNRESET;
+        }
         return -errno;
+    } else if ((size_t)ret < hdr.msg_size) {
+        return -ECONNRESET;
```

Comment:
In case of short socket read/write doing vfu_reset_ctx() will detach. But client(qemu) is not aware of this detach, then wont this will lead to blocking server in attach() again and client will never connect.
Instead should we just assert() in such cases until we fix short socket read/write.

Reply:
I should expand on this in the commit message. The assumption here is that a short read is more likely to be due to the client disconnecting than it is due to a client bug not sending enough data. Thus, it's better to reset rather than blow up.